### PR TITLE
init cosmos-sdk v.0.46.10

### PR DESCRIPTION
**Pull request by satawatnack**
## Implementation details
- update cosmos-sdk version to 0.46.10
- remove `disablefeeless` flag and old oracle ante
- use priority based on gas_prices of uband (MsgReport will have the highest priority) 
- added min-gas-prices on protocol level (hardcoded as 0.0025uband, changed to params in globalfee PR)
- remove legacy migrate cmd
- remove rest-query (added back in fix-rest-query PR)

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,122 +19,154 @@ require (
...
+	google.golang.org/api v0.103.0 // indirect
+	google.golang.org/appengine v1.6.7 // indirect
 	gopkg.in/ini.v1 v1.67.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
 replace (
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
```
Comment (taobun) [inline go.mod:167]: Can we remove this?

File: go.mod [hunk 2]
```diff
@@ -4,10 +4,10 @@ go 1.19
 
 require (
 	github.com/bandprotocol/go-owasm v0.2.3
-	github.com/confio/ics23/go v0.7.0
-	github.com/cosmos/cosmos-sdk v0.45.10
+	github.com/confio/ics23/go v0.9.0
+	github.com/cosmos/cosmos-sdk v0.46.9
```
Comment (taobun) [inline go.mod:8]: Latest tag: out now https://github.com/cosmos/cosmos-sdk/releases/tag/v0.46.10

**Review by taobun — CHANGES_REQUESTED** [review]

File: app/ante.go [hunk 3]
```diff
@@ -35,21 +34,18 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
...
 		ante.NewConsumeGasForTxSizeDecorator(options.AccountKeeper),
-		ante.NewDeductFeeDecorator(options.AccountKeeper, options.BankKeeper, options.FeegrantKeeper),
+		ante.NewDeductFeeDecorator(options.AccountKeeper, options.BankKeeper, options.FeegrantKeeper, options.TxFeeChecker),
 		// SetPubKeyDecorator must be called before all signature verification decorators
 		ante.NewSetPubKeyDecorator(options.AccountKeeper),
 		ante.NewValidateSigCountDecorator(options.AccountKeeper),
 		ante.NewSigGasConsumeDecorator(options.AccountKeeper, sigGasConsumer),
 		ante.NewSigVerificationDecorator(options.AccountKeeper, options.SignModeHandler),
 		ante.NewIncrementSequenceDecorator(options.AccountKeeper),
-		ibcante.NewAnteDecorator(options.IBCKeeper),
```
Comment (taobun) [inline app/ante.go:52]: Why remove IBC ante na?

**Review by taobun — CHANGES_REQUESTED** [review]

File: scripts/start_yoda.sh [hunk 4]
```diff
@@ -9,7 +9,7 @@ yoda config chain-id bandchain
 yoda config validator $(bandd keys show validator -a --bech val --keyring-backend test)
 
 # setup execution endpoint
-yoda config executor "rest:https://iv3lgtv11a.execute-api.ap-southeast-1.amazonaws.com/live/master?timeout=10s"
+yoda config executor "rest:https://asia-southeast2-band-playground.cloudfunctions.net/test-runtime-executor?timeout=10s"
```
Comment (RogerKSI) [inline scripts/start_yoda.sh:12]: please change it to some mock url. e.g. "rest:https://test-runtime-executor.com?timeout=10s"

File: app/app.go [hunk 5]
```diff
@@ -53,8 +51,11 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/genutil"
 	genutiltypes "github.com/cosmos/cosmos-sdk/x/genutil/types"
 	"github.com/cosmos/cosmos-sdk/x/gov"
+	govclient "github.com/cosmos/cosmos-sdk/x/gov/client"
 	govkeeper "github.com/cosmos/cosmos-sdk/x/gov/keeper"
 	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	govv1beta1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
```
Comment (RogerKSI) [inline app/app.go:58]: why need to use both v1 and v1beta1?

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: app/migrate.go [hunk 6]
```diff
@@ -60,150 +38,151 @@ func GenesisDocFromFile(genDocFile string) (*tmtypes.GenesisDoc, error) {
 
 // MigrateGenesisCmd returns a command to execute genesis state migration.
 // nolint: funlen
+// TODO: implement for support cosmos-sdk (v0.46.10)
 func MigrateGenesisCmd() *cobra.Command {
```
Comment (RogerKSI) [inline app/migrate.go:42]: @taobun Hi, want to seek confirmation about this migrate command. Could we remove it?

File: app/app.go (hunk 5 above)
Comment (satawatnack) [reply app/app.go:58]: we need some types from v1beta1

File: app/app.go [hunk 7]
```diff
@@ -722,20 +726,27 @@ func (app *BandApp) GetSubspace(moduleName string) paramstypes.Subspace {
...
 	// Register legacy and grpc-gateway routes for all modules.
-	ModuleBasics.RegisterRESTRoutes(clientCtx, apiSvr.Router)
+	// // removed by cosmos-sdk (v0.46.10)
+	// ModuleBasics.RegisterRESTRoutes(clientCtx, apiSvr.Router)
+
 	ModuleBasics.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
 
 	// Register BandChain rpc routes
-	bandclient.RegisterRoutes(clientCtx, apiSvr.Router)
+	// // removed by cosmos-sdk (v0.46.10)
```
Comment (RogerKSI) [inline app/app.go:748]: remove these comments.

File: app/migrate.go (hunk 6 above)
Comment (taobun) [reply app/migrate.go:42]: Just remove it (legacy migrate method)

Comment (RogerKSI) [reply app/migrate.go:42]: thanks.

**Review by RogerKSI — APPROVED** [review]

File: app/fee/fee_util.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	var priority int64
+	for _, c := range fee {
+		p := int64(math.MaxInt64)
+		// multiplied by 1000 first because priority is int64.
+		// otherwise, if gas_price < 1, the priority will be 0.
+		gasPrice := c.Amount.MulRaw(1000).QuoRaw(gas)
+		if gasPrice.IsInt64() {
+			p = gasPrice.Int64()
+		}
+		if priority == 0 || p < priority {
```
Comment (colmazia) [inline app/fee/fee_util.go:28]: if there 0 coin or <0.001 coin the priority will be zero which will be set again

Comment (RogerKSI) [reply app/fee/fee_util.go]: True, already updated the logic to use priority based on gas price of uband unit.

File: x/oracle/client/common/proof/request_count.go [hunk 9]
```diff
@@ -1,114 +1,114 @@
...
-			Proof: CountProof{
-				BlockHeight:     uint64(commit.Height),
-				CountProof:      requestsCountProof,
-				BlockRelayProof: blockRelay,
-			},
-			EVMProofBytes: evmProofBytes,
-		})
-	}
-}
+// import (
```
Comment (taobun) [inline x/oracle/client/common/proof/request_count.go:3]: Why commend?

File: x/oracle/ibc_module.go [hunk 10]
```diff
@@ -195,15 +196,15 @@ func (im IBCModule) OnRecvPacket(
 ) ibcexported.Acknowledgement {
 	var data types.OracleRequestPacketData
 	if !im.keeper.IBCRequestEnabled(ctx) {
-		return channeltypes.NewErrorAcknowledgement(types.ErrIBCRequestDisabled.Error())
+		return channeltypes.NewErrorAcknowledgement(types.ErrIBCRequestDisabled)
 	} else if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
 		// TODO: Ack with non-deterministic error break consensus?
-		return channeltypes.NewErrorAcknowledgement("cannot unmarshal oracle request packet data")
+		return channeltypes.NewErrorAcknowledgement(fmt.Errorf("cannot unmarshal oracle request packet data"))
```
Comment (taobun) [inline x/oracle/ibc_module.go:202]: Use errors.New(...)

File: x/oracle/module.go [hunk 11]
```diff
@@ -14,11 +13,12 @@ import (
 	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
-	porttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
+
+	// "github.com/cosmos/cosmos-sdk/x/gov/client/rest"
```
Comment (taobun) [inline x/oracle/module.go:17]: Why commend this?

**Review by taobun — CHANGES_REQUESTED** [review]

File: app/fee/fee_util.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	allValidReportMsg := true
+	for _, m := range msgs {
+		r, ok := m.(*types.MsgReportData)
+		// If this is not report msg, skip other msgs on this exec msg
+		if !ok {
+			allValidReportMsg = false
```
Comment (colmazia) [inline app/fee/fee_util.go:138]: No need for this, just return false, nil

Comment (RogerKSI) [reply app/fee/fee_util.go:138]: it fixed in https://github.com/bandprotocol/chain/pull/288/files.

File: go.mod [hunk 13]
```diff
@@ -4,10 +4,10 @@ go 1.19
 
 require (
 	github.com/bandprotocol/go-owasm v0.2.3
-	github.com/confio/ics23/go v0.7.0
-	github.com/cosmos/cosmos-sdk v0.45.10
+	github.com/confio/ics23/go v0.9.0
+	github.com/cosmos/cosmos-sdk v0.46.10
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply go.mod:8]: This will be updated later after all PR is merged.

File: x/oracle/client/common/proof/request_count.go (hunk 9 above)
Comment (RogerKSI) [reply x/oracle/client/common/proof/request_count.go:3]: There is some dependency and function that is no longer usable. However, we moved the rest to use `RegiserNodeService` in this PR (https://github.com/bandprotocol/chain/pull/289)

File: x/oracle/module.go (hunk 11 above)
Comment (RogerKSI) [reply x/oracle/module.go]: removed.

File: app/fee/feechecker.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+			if err := checkValidReportMsg(ctx, fc.oracleKeeper, dr); err != nil {
+				return false, err
+			}
+		} else {
+			isValid, err := checkExecMsgReportFromReporter(ctx, fc.oracleKeeper, msg)
+			if err != nil {
+				return false, err
+			}
+
+			isValidReportTx = isValid
```
Comment (colmazia) [inline app/fee/feechecker.go:89]: bug

Comment (colmazia) [reply app/fee/feechecker.go:89]: but I see it fix in global fee branch

File: x/oracle/ibc_module.go (hunk 10 above)
Comment (RogerKSI) [reply x/oracle/ibc_module.go]: fixed

File: app/fee/feechecker.go (hunk 14 above)
Comment (RogerKSI) [reply app/fee/feechecker.go]: already backport to fix in this PR so that we can review easier.

File: app/fee/feechecker.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func (fc FeeChecker) getGlobalFee(ctx sdk.Context, feeTx sdk.FeeTx) sdk.Coins {
+	gas := feeTx.GetGas()
+
+	glDec := sdk.NewDec(int64(gas))
+	fee := ConsensusMinFee.Mul(glDec)
+	requiredGlobalFees := sdk.NewCoins(sdk.NewCoin(Denom, fee.Ceil().RoundInt()))
+
+	return requiredGlobalFees.Sort()
```
Comment (satawatnack) [inline app/fee/feechecker.go:103]: Why need to sort only one value?

File: app/app.go [hunk 16]
```diff
@@ -465,7 +475,7 @@ func NewBandApp(
 			app.interfaceRegistry,
 		),
 		gov.NewAppModule(appCodec, app.GovKeeper, app.AccountKeeper, app.BankKeeper),
-		mint.NewAppModule(appCodec, app.MintKeeper, app.AccountKeeper),
+		mint.NewAppModule(appCodec, app.MintKeeper, app.AccountKeeper, nil),
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply app/app.go:478]: It will use default value from mint module.

File: app/fee/feechecker.go (hunk 15 above)
Comment (RogerKSI) [reply app/fee/feechecker.go:103]: actually, it's sdk.Coins. So, it's possible to have more than one value.
Although, we hardcoded it for now but we will move it to use params instead in the globalfee PR.

Comment (colmazia) [reply app/fee/feechecker.go:103]: nun si

**Review by RogerKSI — APPROVED** [review]

File: app/genesis.go [hunk 17]
```diff
@@ -69,6 +74,35 @@ func NewDefaultGenesisState() GenesisState {
 	slashingGenesis.Params.DowntimeJailDuration = 60 * 10 * time.Second       // 10 minutes
 	slashingGenesis.Params.SlashFractionDoubleSign = sdk.NewDecWithPrec(5, 2) // 5%
 	slashingGenesis.Params.SlashFractionDowntime = sdk.NewDecWithPrec(1, 4)   // 0.01%
+
+	icaGenesis.HostGenesisState.Params = icahosttypes.Params{
+		HostEnabled: true,
+		AllowMessages: []string{
+			sdk.MsgTypeURL(&authz.MsgExec{}),
```
Comment (taobun) [inline app/genesis.go:81]: Note: Should we allow our message module too?

Comment (RogerKSI) [reply app/genesis.go:81]: @colmazia Please note to add our oracle meesages to upgradeHandler 2.5

File: app/ante.go [hunk 18]
```diff
@@ -34,22 +41,31 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
...
+	if options.TxFeeChecker == nil {
+		feeChecker := fee.NewFeeChecker(options.OracleKeeper)
+		options.TxFeeChecker = feeChecker.CheckTxFeeWithMinGasPrices
+	}
+
 	anteDecorators := []sdk.AnteDecorator{
-		ante.NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
-		ante.NewRejectExtensionOptionsDecorator(),
-		ante.NewMempoolFeeDecorator(),
+		ante.NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called firs(),
```
Comment (taobun) [inline app/ante.go:50]: first?

Comment (RogerKSI) [reply app/ante.go]: fixed

File: app/genesis.go (hunk 17 above)
Comment (RogerKSI) [reply app/genesis.go:81]: It will be added in the upgradeHandler 2.5 PR.

**Review by taobun — APPROVED** [review]

**Pull request merged**